src/connection.ts

```diff
@@ -3718,54 +3713,46 @@ Connection.prototype.STATE = {
-              const getTokenFromCredentials = (err: Error | undefined, credentials?: UserTokenCredentials | MSIAppServiceTokenCredentials | MSIVmTokenCredentials | ApplicationTokenCredentials) => {
-                if (err) {
-                  return callback(err);
-                }
-
-                credentials!.getToken().then((tokenResponse: { accessToken: string | undefined }) => {
-                  callback(null, tokenResponse.accessToken);
-                }, callback);
-              };
+              const getTokenFromCredentials = (credentials: TokenCredential) =>
+                credentials.getToken([ tokenScope ]).then(
+                  (tokenResponse) => callback(null, tokenResponse?.token),
+                  callback
+                );
 
               if (authentication.type === 'azure-active-directory-password') {
-                const credentials = new UserTokenCredentials(
-                  '7f98cb04-cd1e-40df-9140-3bf7e2cea4db',
-                  authentication.options.domain ?? 'common',
+                const credentials = new UsernamePasswordCredential(
+                  authentication.options.domain ?? 'common',  // tenantId
+                  '7f98cb04-cd1e-40df-9140-3bf7e2cea4db',     // clientId
                   authentication.options.userName,
-                  authentication.options.password,
-                  fedAuthInfoToken.spn,
-                  undefined, // environment
-                  authenticationCache
+                  authentication.options.password
                 );
 
-                getTokenFromCredentials(undefined, credentials);
-              } else if (authentication.type === 'azure-active-directory-msi-vm') {
-                loginWithVmMSI({
-                  clientId: authentication.options.clientId,
-                  msiEndpoint: authentication.options.msiEndpoint,
-                  resource: fedAuthInfoToken.spn
-                }, getTokenFromCredentials);
-              } else if (authentication.type === 'azure-active-directory-msi-app-service') {
-                loginWithAppServiceMSI({
-                  msiEndpoint: authentication.options.msiEndpoint,
-                  msiSecret: authentication.options.msiSecret,
-                  resource: fedAuthInfoToken.spn,
-                  clientId: authentication.options.clientId
-                }, getTokenFromCredentials);
+                getTokenFromCredentials(credentials);
+              } else if (authentication.type === 'azure-active-directory-msi-vm' || authentication.type === 'azure-active-directory-msi-app-service') {
+                const msiArgs = authentication.options.clientId ? [ authentication.options.clientId, {} ] : [ {} ];
+                const credentials = new ManagedIdentityCredential(...msiArgs);
+
+                if (authentication.options.msiEndpoint) {
+                  process.env.IDENTITY_ENDPOINT = authentication.options.msiEndpoint;
```

Comment:
Note to reviewers: I don't like that this assigns values to `process.env`, but I couldn't find any alternatives in the `@azure/identity` options...

https://github.com/Azure/azure-sdk-for-js/blob/0596e3a57f84cd3289c4bdad1271da7aa1f8eb37/sdk/identity/identity/src/credentials/managedIdentityCredential/arcMsi.ts#L86

https://github.com/Azure/azure-sdk-for-js/blob/0596e3a57f84cd3289c4bdad1271da7aa1f8eb37/sdk/identity/identity/src/credentials/managedIdentityCredential/appServiceMsi2017.ts#L51-L52

Reply:
Should we maybe just deprecate and remove the `msiEndpoint` and  `msiSecret` options? What do you think?

Reply:
Yeah I was thinking the same.

Would you want to keep the changes and just add some `depd` logs + documentation updates? I can take a look at adding those updates if so.

Reply:
I think we should prepare a PR that will deprecate explicitly setting these options (and make them fallback to the environment variables if not set explicitly), and release that first.

And then mark these changes here as breaking and release them with a new major version bump.